### DISCUSSION TITLE
Problem with --js-library, dynamic linking and function in dynamic library

### DISCUSSION THREAD

**Discussion by Stuonts** [emscripten-core/emscripten#16997] [Q&A] [answered]
We have a project that needs dynamic linking and may use Javascript functions in dynamic libraries
In case if function from JS library is used in dynamic library, it gives error
"Uncaught (in promise) RuntimeError: Aborted(Assertion failed: Missing signature argument to addFunction:"
Generated dynamic library contains 'GOT.functestJS'
Generated javascript traverses GOT dictionary in reportUndefinedSymbols and tries to use .sig (which is absent)

Seems that either testJS function shouldn't be marked as GOT.func in .wasm or .sig should generated in JS for testJS
Also, some build flags can cause the issue. Please advice.

Emscripten version: 3.1.8
Attached sample (build for Windows in bat, 'set EMSDK_DIR=F:\.emsdk\upstream\emscripten' should be replaced in build.bat
[sig.zip](https://github.com/emscripten-core/emscripten/files/8739952/sig.zip)

**Comment by sbc100** [comment, marked as answer]
In order to JS functions to be address taken you need to have `test_JS__sig: 'v'`  to you library JS file.   The signature (in this case 'v') is a sequence of characters representing the first the return code then that arguments.   So `vi` would a void function that takes a single int.

**Reply by Stuonts** [reply]
so, Emscripten should generate .sig in Javascript file? Is it a bug of Emscripten?

**Reply by sbc100** [reply]
No, you (currently) need to explicitly write the `__sig` into your library file.  Kind of like we do in the core libraries. e.g:

https://github.com/emscripten-core/emscripten/blob/1314e69c8d9aa22f2109e6f94734a972476f0376/src/library.js#L31

**Reply by Stuonts** [reply]
@sbc100 ,
we also have the same issue with library functions, for example
with emval_new
There is no emval_new__sig in emval.js while there are __sig for some other functions (not all)

**Reply by sbc100** [reply]
Yes, we tent to add those `__sig` attributes as needed.  Feel free to set a PR at add more if you need them.

**Reply by Stuonts** [reply]
@sbc100 
https://github.com/emscripten-core/emscripten/pull/17075
